File: benchmarks/compare/raptor/python/raptor/journey_factory.py
"""JourneyFactory (mirrors src/results/JourneyFactory.ts).

Extracts Journey[] from the kConnections index for a destination. The per-
destination round map is keyed by integer k and must be iterated in NUMERIC
ASCENDING order (contract #1 exception).
"""
from __future__ import annotations

from typing import List

from .gtfs import StopID, Time, TimetableLeg, Transfer
from .journey import AnyLeg, Journey, is_transfer
from .scan_results import ConnectionIndex


class JourneyFactory:
    def getResults(self, k_connections: ConnectionIndex, destination: StopID) -> List[Journey]:
        results: List[Journey] = []

        dest_map = k_connections.get(destination) or {}
        # Round numbers are integer-like keys -> iterate numeric ascending.
        for k in sorted(dest_map, key=lambda x: int(x)):
            legs = self._get_journey_legs(k_connections, k, destination)
            departure_time = self._get_departure_time(legs)
            arrival_time = self._get_arrival_time(legs)

            results.append(Journey(legs=legs, departureTime=departure_time, arrivalTime=arrival_time))

        return results

    def _get_journey_legs(
        self, k_connections: ConnectionIndex, k, final_destination: StopID
    ) -> List[AnyLeg]:
        legs: List[AnyLeg] = []

        destination = final_destination
        i = int(k)
        while i > 0:
            connection = k_connections[destination][i]

            if is_transfer(connection):
                legs.append(connection)
                destination = connection.origin
            else:
                trip, start, end = connection
                stop_times = trip.stopTimes[start:end + 1]
                origin = stop_times[0].stop

                legs.append(
                    TimetableLeg(
                        stopTimes=stop_times,
                        origin=origin,
                        destination=destination,
                        trip=trip,
                    )
                )
                destination = origin

            i -= 1

        legs.reverse()
        return legs

    def _get_departure_time(self, legs: List[AnyLeg]) -> Time:
        transfer_duration = 0

        for leg in legs:
            if not self._is_timetable_leg(leg):
                transfer_duration += leg.duration
            else:
                return leg.stopTimes[0].departureTime - transfer_duration

        return 0

    def _get_arrival_time(self, legs: List[AnyLeg]) -> Time:
        transfer_duration = 0

        for leg in reversed(legs):
            if not self._is_timetable_leg(leg):
                transfer_duration += leg.duration
            else:
                return leg.stopTimes[-1].arrivalTime + transfer_duration

        return 0

    def _is_timetable_leg(self, connection: AnyLeg) -> bool:
        return isinstance(connection, TimetableLeg)

File: benchmarks/compare/raptor/node/src/results/JourneyFactory.js
import { isTransfer } from "./ResultsFactory.js";

/**
 * Extracts journeys from the kConnections index.
 */
export class JourneyFactory {
  /**
   * Take the best result of each round for the given destination and turn it into a journey.
   */
  getResults(kConnections, destination) {
    const results = [];

    // Object.keys over an integer-keyed object yields keys in numeric-ascending order.
    for (const k of Object.keys(kConnections[destination] || {})) {
      const legs = this.getJourneyLegs(kConnections, k, destination);
      const departureTime = this.getDepartureTime(legs);
      const arrivalTime = this.getArrivalTime(legs);

      results.push({ legs, departureTime, arrivalTime });
    }

    return results;
  }

  /**
   * Iterate back through each connection and build up a series of legs to plan the journey
   */
  getJourneyLegs(kConnections, k, finalDestination) {
    const legs = [];

    for (let destination = finalDestination, i = parseInt(k, 10); i > 0; i--) {
      const connection = kConnections[destination][i];

      if (isTransfer(connection)) {
        legs.push(connection);

        destination = connection.origin;
      } else {
        const [trip, start, end] = connection;
        const stopTimes = trip.stopTimes.slice(start, end + 1);
        const origin = stopTimes[0].stop;

        legs.push({ stopTimes, origin, destination, trip });

        destination = origin;
      }
    }

    return legs.reverse();
  }

  getDepartureTime(legs) {
    let transferDuration = 0;

    for (const leg of legs) {
      if (!this.isTimetableLeg(leg)) {
        transferDuration += leg.duration;
      }
      else {
        return leg.stopTimes[0].departureTime - transferDuration;
      }
    }

    return 0;
  }

  getArrivalTime(legs) {
    let transferDuration = 0;

    for (let i = legs.length - 1; i >= 0; i--) {
      const leg = legs[i];

      if (!this.isTimetableLeg(leg)) {
        transferDuration += leg.duration;
      }
      else {
        return leg.stopTimes[leg.stopTimes.length - 1].arrivalTime + transferDuration;
      }
    }

    return 0;
  }

  isTimetableLeg(connection) {
    return connection.stopTimes !== undefined;
  }
}
